### PR TITLE
Implement partial signing

### PR DESCRIPTION
**Pull request by GNUp**

### CODE REVIEW

**Comment by kseo** [comment]
@GNUp The Travis CI build failed. Please fix it first.

File: types/src/transaction/transaction.rs [hunk 1]
```diff
@@ -264,6 +268,77 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+    mut result: Vec<AssetTransferOutput>,
+) -> Vec<AssetTransferOutput> {
+    match bitmask {
+        0 => return result,
+        _ => {
+            if ((bitmask & 0x1) == 1) && !outputs.is_empty() {
+                result.push(outputs.remove(0));
+                apply_bitmask(bitmask >> 1, outputs, result)
+            } else {
+                apply_bitmask(bitmask >> 1, outputs, result)
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: I think you must remove output here.

File: types/src/transaction/transaction.rs [hunk 2]
```diff
@@ -264,6 +268,77 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
 
+fn apply_bitmask(
+    bitmask: u128,
+    mut outputs: Vec<AssetTransferOutput>,
+    mut result: Vec<AssetTransferOutput>,
+) -> Vec<AssetTransferOutput> {
+    match bitmask {
+        0 => return result,
+        _ => {
+            if ((bitmask & 0x1) == 1) && !outputs.is_empty() {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: I think it must be the error if the bit is set but the corresponding output does not exist.

File: vm/src/executor.rs [hunk 3]
```diff
@@ -139,12 +140,24 @@ impl Stack {
     }
 }
 
-pub fn execute(
+fn make_bitmask(mut item: Vec<u8>, mut result: u128) -> u128 {
+    if item.is_empty() {
+        return result
+    } else {
```
Comment (sgkim126) [inline vm/src/executor.rs]: Do not use `else` when the above statements always return.

File: types/src/transaction/transaction.rs [hunk 4]
```diff
@@ -264,6 +268,77 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+                result.push(outputs.remove(0));
+                apply_bitmask(bitmask >> 1, outputs, result)
+            } else {
+                apply_bitmask(bitmask >> 1, outputs, result)
+            }
+        }
+    }
+}
+
+impl<'a> PartialHashing for &'a Transaction {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: Why do you implement `PartialHashing` for the reference of `Transaction` instead of `Transaction`?

File: types/src/transaction/transaction.rs [hunk 5]
```diff
@@ -264,6 +268,77 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+
+        match self {
+            Transaction::AssetTransfer {
+                network_id,
+                burns,
+                inputs,
+                outputs,
+                nonce,
+            } => {
+                let new_burns: Vec<_> = if input_scheme {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: I think rust compiler can infer that this variable is a `Vec`.

File: types/src/transaction/transaction.rs (hunk 4 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: Cause when `execute` is called, there is only a borrowed transaction we can use in `fn transfer_asset`.

File: vm/src/executor.rs [hunk 6]
```diff
@@ -16,10 +16,11 @@
 
 use ccrypto::{blake256, keccak256, ripemd160, sha256};
 use ckey::{verify, Public, Signature, SIGNATURE_LENGTH};
-use primitives::H256;
+use ctypes::transaction::{AssetOutPoint, HashingError, PartialHashing};
 
 use instruction::{has_expensive_opcodes, is_valid_unlock_script, Instruction};
 
+
```
Comment (joojis) [inline vm/src/executor.rs]: Don't include unnecessary change.

Comment (GNUp) [reply vm/src/executor.rs]: Okay, I will pay more attention next time.

File: types/src/transaction/transaction.rs [hunk 7]
```diff
@@ -264,6 +273,81 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+    mut outputs: Vec<AssetTransferOutput>,
+    mut result: Vec<AssetTransferOutput>,
+) -> Result<Vec<AssetTransferOutput>, HashingError> {
+    match bitmask {
+        0 => return Ok(result),
+        _ => {
+            if outputs.is_empty() {
+                return Err(HashingError::InvalidFilter)
+            }
+            if (bitmask & 0x1) == 1 {
```
[the remark this answers is not in the document]
Comment (sgkim126) [reply types/src/transaction/transaction.rs]: FYI, it's bad for performance to remove the first item of a vector.

**Comment by sgkim126** [comment]
Please add some test cases about partial signing.

File: types/src/transaction/transaction.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (GNUp) [reply types/src/transaction/transaction.rs]: Okay, Thank you

File: vm/src/executor.rs [hunk 8]
```diff
@@ -47,6 +47,13 @@ pub enum RuntimeError {
     IndexOutOfBound,
     StackUnderflow,
     TypeMismatch,
+    InvalidFilter,
+}
+
+impl From<HashingError> for RuntimeError {
+    fn from(_error: HashingError) -> Self {
+        RuntimeError::InvalidFilter
```
[the remark this answers is not in the document]
Comment (GNUp) [reply vm/src/executor.rs]: Okay, I didn't think about it.

File: types/src/transaction/transaction.rs [hunk 9]
```diff
@@ -264,6 +232,148 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+    for i in 0.. {
+        if bitmask == 0 {
+            return Ok(result)
+        }
+
+        if outputs.is_empty() {
+            return Err(HashingError::InvalidFilter)
+        }
+
+        let mut output = outputs[i].clone();
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: 1. No reason to make `output` mutable.
2. No reason to clone `outputs[i]` if it's not pushed in the result.

```rust
if (bitmask & 0x1) == 1 {
    result.push(outputs[i].clone());
}
```

File: types/src/transaction/transaction.rs [hunk 10]
```diff
@@ -264,6 +232,148 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+        match self {
+            Transaction::AssetTransfer {
+                network_id,
+                burns,
+                inputs,
+                outputs,
+                nonce,
+            } => {
+                let new_burns = if input_scheme {
+                    let mut tmp: Vec<_> = burns
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: Please use proper name instead of `tmp`

File: types/src/transaction/transaction.rs (hunk 9 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I resolved it

File: types/src/transaction/transaction.rs (hunk 10 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I resolved it

File: types/src/transaction/transaction.rs [hunk 11]
```diff
@@ -754,4 +862,22 @@ mod tests {
...
+            outputs: Vec::new(),
+            nonce: 0,
+        };
+        let outpoint = AssetOutPoint {
+            transaction_hash: H256::default(),
+            index: 0,
+            asset_type: H256::default(),
+            amount: 0,
+        };
+        assert_eq!(transaction.hash_partially(5 as u128, &outpoint, false), Err(HashingError::InvalidFilter));
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: `0b101` instead of 5.
Using binary format is better since it's a bitmask.

Comment (GNUp) [reply types/src/transaction/transaction.rs]: I resolved it

File: types/src/transaction/transaction.rs [hunk 12]
```diff
@@ -264,6 +232,146 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+            result.push(outputs[i].clone());
+        }
+        bitmask = bitmask >> 1
+    }
+    Err(HashingError::InvalidFilter)
+}
+
+impl PartialHashing for Transaction {
+    fn hash_partially(&self, bitmask: u128, cur: &AssetOutPoint, burn: bool) -> Result<H256, HashingError> {
+        let input_scheme = (bitmask & 0x1) == 1;
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: `let sign_all_input = ...`

File: types/src/transaction/transaction.rs [hunk 13]
```diff
@@ -264,6 +232,146 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+        let output_scheme = (bitmask >> 1 & 0x1) == 1;
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: `let sign_all_output = ...`

File: types/src/transaction/transaction.rs [hunk 14]
```diff
@@ -264,6 +232,146 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+
+        match self {
+            Transaction::AssetTransfer {
+                network_id,
+                burns,
+                inputs,
+                outputs,
+                nonce,
+            } => {
+                let new_burns = if input_scheme {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: It seems that 271~303 and 305~337 do the same thing.
Please extract it a function.

File: types/src/transaction/transaction.rs [hunk 15]
```diff
@@ -264,6 +232,146 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+                    let mut burns_without_script: Vec<_> = burns
+                        .iter()
+                        .map(|input| AssetTransferInput {
+                            prev_out: input.prev_out.clone(),
+                            lock_script: Vec::new(),
+                            unlock_script: Vec::new(),
+                        })
+                        .collect();
+                    burns_without_script.push(AssetTransferInput {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: IMO, appending one item for the signing all scheme seems too ad-hoc.
How about passing a bitmask to the blake function?

File: types/src/transaction/transaction.rs (hunk 12 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I changed it

File: types/src/transaction/transaction.rs (hunk 13 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I changed it

File: types/src/transaction/transaction.rs (hunk 15 above)
Comment (sgkim126) [reply types/src/transaction/transaction.rs]: FYI, There is a function named [blake256_with_key](https://github.com/CodeChain-io/codechain/blob/049e64991339fc5ec85bb0efcc08cbcb16993d08/crypto/src/blake.rs#L26)

File: types/src/transaction/transaction.rs (hunk 14 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I resolved it

File: types/src/transaction/transaction.rs (hunk 15 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I resolved it

File: types/src/transaction/transaction.rs [hunk 16]
```diff
@@ -264,6 +232,115 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+            result.push(outputs[i].clone());
+        }
+        bitmask = bitmask >> 1
+    }
+    Err(HashingError::InvalidFilter)
+}
+
+fn apply_sign_scheme_to_input(
+    inputs: &Vec<AssetTransferInput>,
+    scheme: bool,
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: Rename scheme to is_sign_all

File: types/src/transaction/transaction.rs [hunk 17]
```diff
@@ -264,6 +232,115 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+            .iter()
+            .map(|input| AssetTransferInput {
+                prev_out: input.prev_out.clone(),
+                lock_script: Vec::new(),
+                unlock_script: Vec::new(),
+            })
+            .collect()
+    }
+
+    if is_burn {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: Would you explain what it is?

File: types/src/transaction/transaction.rs [hunk 18]
```diff
@@ -264,6 +232,115 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+                outputs,
+                nonce,
+            } => {
+                let new_burns = apply_sign_scheme_to_input(burns, sign_all_inputs, burn, cur);
+                let new_inputs = apply_sign_scheme_to_input(inputs, sign_all_inputs, burn, cur);
+
+                let new_outputs = if sign_all_outputs {
+                    outputs
+                        .iter()
+                        .map(|output| AssetTransferOutput {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: Why not `.map(Clone::clone)`?

Comment (sgkim126) [reply types/src/transaction/transaction.rs]: It seems that it can be `outputs.clone()` if it can be `.map(Clone::clone).

File: types/src/transaction/transaction.rs [hunk 19]
```diff
@@ -264,6 +232,115 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
     sum.iter().all(|(_, sum)| sum.is_zero())
 }
 
+fn make_tag(mut item: Vec<u8>, mut result: u128) -> u128 {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: IMO, passing a vector reference and an index instead of a mutable vector will be better.

Comment (sgkim126) [reply types/src/transaction/transaction.rs]: For instance, you can refactor this function like the below.
```rust
fn make_tag(item: &[u8]) -> u128 {
    if item.is_empty() {
        return 0
    }
    fn make_tag_internal(item: &[u8], index: usize, mut result: u128) -> u128 {
        if item.len() == index {
            return result
        }
        // Assume that the mask is stored as little endian
        let head = item[index] as u128;
        result = (result << 8) | head as u128;
        make_tag_internal(item, index + 1, result)
    }
    make_tag_internal(item, 0, 0)
}
```

File: types/src/transaction/transaction.rs [hunk 20]
```diff
@@ -264,6 +232,115 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+    if item.is_empty() {
+        return result
+    }
+    // Assume that the mask is stored as little endian
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: Usually, a bitmask uses a big-endian. Why do you use a little-endian?

File: types/src/transaction/transaction.rs (hunk 16 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I changed it

File: types/src/transaction/transaction.rs (hunk 18 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: It is my mistake. I didn't recover that from the old scheme. In the previous version I implemented, there was an additional element for storing filter in output and input. So I had to clone other elements one by one like this way. However, there is no need to do this now.

File: types/src/transaction/transaction.rs (hunk 19 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: It also seems more elegant because the function's arguments are more intuitive. I replaced it with the code you provided.

File: types/src/transaction/transaction.rs (hunk 20 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I just specified the way how u128, the bitmask, is stored in Vec<u8>. I did it cause it was a more familiar way. I change it to stored as big-endian.

File: types/src/transaction/transaction.rs [hunk 21]
```diff
@@ -264,6 +232,113 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
     sum.iter().all(|(_, sum)| sum.is_zero())
 }
 
+fn make_tag(item: &[u8]) -> u128 {
+    if item.is_empty() {
+        return 0
+    }
+    debug_assert!(item.len() == 16, "The size of filter vec should be 16");
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]:
```rust
debug_assert_eq!(16, item.len());
```

File: state/src/impls/shard_level.rs [hunk 22]
```diff
@@ -159,7 +159,12 @@ impl<B: Backend + ShardBackend> ShardLevelState<B> {
                 inputs,
                 outputs,
                 ..
-            } => self.transfer_asset(&transaction, sender, burns, inputs, outputs),
+            } => {
+                if outputs.len() > 126 {
```
Comment (sgkim126) [inline state/src/impls/shard_level.rs]: I think it should be an assertion and verify method should fail.

File: types/src/transaction/transaction.rs (hunk 21 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I resolved it

File: state/src/impls/shard_level.rs (hunk 22 above)
Comment (GNUp) [reply state/src/impls/shard_level.rs]: I moved it to verify method

File: types/src/transaction/transaction.rs [hunk 23]
```diff
@@ -224,6 +192,9 @@ impl Transaction {
                 outputs,
                 ..
             } => {
+                if outputs.len() > 26 {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: 126

**Comment by kseo** [comment]
@GNUp Please resolve conflicts.

**Comment by joojis** [comment]
@GNUp Please update the SDK version to `0.4.0-rc1` for the integration tests

File: types/src/transaction/transaction.rs (hunk 16 above)
Comment (sgkim126) [reply types/src/transaction/transaction.rs]: It's not changed yet.

File: types/src/transaction/transaction.rs [hunk 24]
```diff
@@ -346,6 +317,109 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+
+    loop {
+        if !(index < filter_len) {
+            return Ok(result)
+        }
+
+        let mut filter = bitmask.pop().unwrap();
+
+        for i in 0..8 {
+            if index == output_len as usize {
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: Move it out of for loop.

File: types/src/transaction/transaction.rs [hunk 25]
```diff
@@ -346,6 +317,109 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
     sum.iter().all(|(_, sum)| sum.is_zero())
 }
 
+fn apply_bitmask_to_output(
+    mut bitmask: Vec<u8>,
+    filter_len: usize,
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: Why does it need `filter_len`?
Isn't it the length of `bitmask`?

File: types/src/transaction/transaction.rs [hunk 26]
```diff
@@ -346,6 +317,109 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+                let new_burns = is_sign_all(burns, sign_all_inputs, is_burn, cur);
+                let new_inputs = is_sign_all(inputs, sign_all_inputs, !is_burn, cur);
+
+                let new_outputs = if sign_all_outputs {
+                    outputs.clone()
+                } else {
+                    if bitvec.len() != filter_len as usize {
+                        return Err(HashingError::InvalidFilter)
+                    }
+                    apply_bitmask_to_output(bitvec.clone(), filter_len as usize, outputs.clone(), Vec::new())?
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: What happens if the length of `outputs` is not the same with `bitvec.len()`

File: types/src/transaction/transaction.rs [hunk 27]
```diff
@@ -346,6 +317,109 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+                result.push(outputs[index + i].clone());
+            }
+
+            filter = filter >> 1;
+        }
+        index += 1;
+    }
+}
+
+fn is_sign_all(
```
Comment (sgkim126) [inline types/src/transaction/transaction.rs]: IMO, the name of this function is not matched with this behavior.

File: types/src/transaction/transaction.rs (hunk 24 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: It is actually (index + i). my mistake, I changed it.

File: types/src/transaction/transaction.rs (hunk 25 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I removed that argument.

File: types/src/transaction/transaction.rs (hunk 26 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I defined some rules.

1. filter_len must be same with the bitvec's byte length which is the tag except for one byte for output scheme, input scheme and filter_len.
2. filter_len can be greater than output's length. Only exist outputs will be signed. So an output can be added if a signer wants to do. He just adds zero behind the filter. If he doesn't want to extend the outputs, he can add 1 behind the filter.
3. filter_len can be shorter than output's length. Only relevant outputs will be signed.

I am now writing the specification for partial signing. If you give me your opinion, I will summarize that with the above rules.

File: types/src/transaction/transaction.rs (hunk 16 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I misunderstand your request. I changed it.

File: types/src/transaction/transaction.rs (hunk 27 above)
Comment (GNUp) [reply types/src/transaction/transaction.rs]: I changed it to `apply_input_scheme`

File: types/src/transaction/transaction.rs [hunk 28]
```diff
@@ -346,6 +317,107 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
...
+fn apply_bitmask_to_output(
+    mut bitmask: Vec<u8>,
+    outputs: Vec<AssetTransferOutput>,
+    mut result: Vec<AssetTransferOutput>,
+) -> Result<Vec<AssetTransferOutput>, HashingError> {
+    let mut index = 0;
+    let output_len = outputs.len();
+
+    loop {
+        if !((index as usize) < bitmask.len()) {
```
[the remark this answers is not in the document]
Comment (GNUp) [reply types/src/transaction/transaction.rs]: Oh, I fixed it.

**Comment by joojis** [comment]
Let's merge these patches at here https://github.com/CodeChain-io/codechain/pull/789

**Comment by GNUp** [comment]
@joojis Okay, I will close this PR. @sgkim126
